trin-core/src/portalnet/protocol.rs

```diff
@@ -182,6 +191,26 @@ impl PortalnetEvents {
 
         Ok(response)
     }
+
+    // This could be handled in the UtpListener impl, but for consistency with data handling
+    // and clarity it can be put here
+    async fn process_utp_byte_stream(&mut self) {
+        for (_, conn) in self.utp_listener.utp_connections.iter_mut() {
+            let received_stream = conn.recv_data_stream.clone();
+            if received_stream.is_empty() {
+                continue;
+            }
+
+            let message_len = BigEndian::read_u32(&received_stream[0..4]);
+
+            if message_len + 4 > received_stream.len() as u32 {
+                // handle data function(&received_stream[4..(message_len + 4)]
+
+                // Removed the handled data from the buffer
+                conn.recv_data_stream = received_stream[((message_len + 4) as usize)..].to_owned();
```

Comment:
> I'm missing something about how this works. On a casual read, it looks like it strips the first four bytes out of every data stream every time a new packet comes in. So some comments to explain what's really going on would help future code readers here.
> 
> By my reading, this 4-byte length prefix is above the uTP transport layer in the protocol stack. So I definitely agree that it doesn't belong in UtpListener

That is what we were going over in the call it is assumed when doing connection.send() for example on your end 

```
let buf = vec![];
BigEndian::write_u32_into(&[msg_transport.len()], buf);
buf.append(msg_transport);
connection.send(buf);
```

You guys need to add a delimiter or something I would assume because you can get 9 bytes in 9 different packets (which you would have to wait for) or 9 bytes in one packet. uTP is just revealing a stream of data to you, if you can't parse it I would assume it is useless to you. Also, you guys would be the ones closing it. So currently it is connection.send(| len | payload |) this can take multiple packets to get to you which is why I would think you would use a delimiter.

| len | payload | is just the delimiter Piper suggested, but you guys could use whatever you want uTP won't be handling that.

### Explaining the impl's in the uTP implementation
uTP should be treated no differently than we would implement something like TCP for example. 

Anyways I guess I will define what UtpListener is and UtpStream is.

#### UtpListener
UtpListener is basically a mailman sorting mail(A talk request) into the various mailboxes (connections) of an apartment. Because we are building uTP over Discv5 it is weird in the sense that we are giving the mailman mail is seen where normally it would be fully abstracted. this mail task is handled in UtpListener::process_utp_request()

![image](https://user-images.githubusercontent.com/31669092/131203181-abec8e3c-98f4-47bf-aa25-d5c0771acb8f.png)
This is the connect feature we use ConnectionKey is like the connections home address in a sense. So utp_connections is just a list of all of your local mailboxs.

#### UtpStream
UtpStream is just an individual connection to another node. This is where all of the code makes sure we get our packets in the right order among other things and reveals the stream. You could read more of this from the uTP spec and this is assuming you have some knowledge of how TCP works.

### Talking about the pseudocode
`msg_transport = UTPListener()`
This would basically be the equivalent of creating a new UDP socket your application would only have one connected to a port (binding to a socket address). Same with TCP you would have one Listener which would handling the incoming packets and sort the packets into the connections or TcpStream (The naming scheme is the same with Rusts TCP lib) My uTP implementation would be used the same way you would use Rust's TCP library (uTP is just TCP built on top of UDP(but in this special exception it is TCP built on Discv5))
![image](https://user-images.githubusercontent.com/31669092/131201600-44f5d0e6-d439-40f4-891b-d5fe1c89efb2.png)

So when I look at this I could see some way it could work, but I also see a ton of issues surrounding that.

First off I have been kind of confused why you guys want to create your own connection ID, when it would only be internally used by uTP and seems kind of redundant in a way, but that is fine since it really doesn't matter at the end of the day.

I was going to do a write-up, but decided to really quickly just do some pseudocode based around what Calver asked for (keep in mind uTP has the same API as TCP, TCP is almost identical to uTP. uTP is just TCP built on top of UDP, but in this case it is built on top of Discv5) I guess this is weird to think about because most of Trin is designed around a Datagram oriented protocol.

Anyways here is my pseudocode, based on the current uTP implementation which has a pretty standard TCP-ish API. This example is assuming we are using Pipers suggested delimiter

```rust
fn main() {
    let utp_listener = UtpListener {
        discovery: discovery.clone(),
        utp_connections: HashMap::new(),
    };

   //We are assuming that we already have discv5 pumping data into the UtpListener

    let connect_id = 421412;
    let node_id: NodeId = "Random node ID";
    utp_listener.connect(connect_id, node_id).await;
    if let Some(conn) = utp_listener.utp_connections.get_mut(&ConnectionKey {
        node_id: node_id.clone(),
        conn_id_recv: connection_id,
    }) {
        send_request_proofs(conn);
    }
}

fn send_request_proofs(connection) {
    request_proofs = RequestProofsMsg();

    // This could either be the conn_id_recv (our id) or conn_id_send (their id) I will assume it is ours, but this seems useless
    request_proofs.connection_id = connection.conn_id_recv;
    let buf = vec![];
    BigEndian::write_u32_into(&[request_proofs.len()], buf);
    buf.append(request_proofs);
    connection.write(buf);

    proof_receiver = ProofReceiver(connection);
    spawn!(handle_proof(proof_receiver).await);
}

async fn handle_proof(proof_receiver) {
    try {
        proof = proof_receiver.receive().await;
    } except BadProof {
        // Wrong length, or otherwise invalid. Log & return
        return;
    } else {
        save_proof(proof).await;
    }
}

struct ProofReceiver {
    connection: stuff
}

impl ProofReceiver {
    // Although I personally wouldn't want to do this way, we are basically dealing
    // with the same contrants we would using TCP.
    async fn how_I_see_this_being_currently_impl_receive(self) {
        loop {
            // We should sleep which is unidle, but because we want to use (basically TCP) uTP like this
            // we have to do weird unoptimal things to not waste clock cycles
            tokio::time::sleep(Duration::from_millis(100)).await;
            let received_stream = self.connection.recv_data_stream.clone();
            if received_stream.is_empty() {
                continue;
            }

            let message_len = BigEndian::read_u32(&received_stream[0..4]);

            if message_len + 4 > received_stream.len() as u32 {
                conn.recv_data_stream = received_stream[((message_len + 4) as usize)..].to_owned();
                return Proof(&received_stream[4..(message_len + 4));
            }
        }
    }

    async fn Calver's_receive(self) {
        let payload = self.connection.receive_all().await;
        let (num_bytes, body) = (to_int(payload[:4]), payload[4:]);
        if len(body) != num_bytes {
            //This makes no sense to me uTP has no way to tell you if you got all of your data so the
            // receive_all thing wouldn't work so I don't get this if statement because it serves
            // as a completely different purpose then in what I supposed process_utp_byte_stream()
            // where it checks if it has all of the data is there if so it handles the data, if it doesn't
            // you leave the data alone and wait till the next time it runs, This isn't a Datagram
            // oriented protocol like UDP, you guys will need to check
            // if you have received all of the bytes you needed I get that it is pseudocode, and
            // you guys all have different ideas how it should work, but this is basically just TCP
            // and shouldn't be treated much differenly.
            raise BadProof("incomplete proof received, likely peer disconnect");
        }
        return Proof(body);  // might raise BadProof
    }
}
```

Normally you could handle your own TcpStreams, but because we are not as low level as TCP we store a connection_list in uTP_listner of connections to pass the packets to.

Anyways I don't really like how this is being done possibly we could change recv_data_stream to a channel() and have the receive function be like this instead

```rust
struct ProofReceiver {
    connection: stuff
    buffer: &[u8]
}

impl ProofReceiver {
    async fn idea_receive(mut self) {
        while let Some(buffer) = self.connection.recv_data_stream.recv().await {
            let mut received_stream = vec![];
            if !buffer.is_empty() {
                received_stream.append(buffer);
            }
            received_stream.append(self.connection.recv_data_stream.clone());
            if received_stream.is_empty() {
                continue;
            }

            let message_len = BigEndian::read_u32(&received_stream[0..4]);

            if message_len + 4 >= received_stream.len() as u32 {
                conn.recv_data_stream = received_stream[((message_len + 4) as usize)..].to_owned();
                return Proof(&received_stream[4..(message_len + 4));
            } else { 
                self.buffer = received_stream;
            }
        }
    }
}
```
well this is cleaner then before saying we are not using sleep, I would say this solution isn't very idle either

Originally I actually did want to give you all of the data back when it was all received, but Piper said uTP shouldn't have to handle that and honestly I ended up agreeing with him.

Reply:
> Also, you guys would be the ones closing it

Who is "you guys" here?

> First off I have been kind of confused why you guys want to create your own connection ID, when it would only be internally used by uTP and seems kind of redundant in a way,

Heh, this time I understand "you guys", but also it suggests that there are some people chiseling specs into stone. We are just a loosely assembled collective, building out a draft spec together. You're _in_ as much as anyone. So you can use "we" instead to remind yourself that we're all just figuring it out together. :)

So, I'll bite: what's another way to correctly deliver an incoming utp packet, when there are two streams opened with the same peer?

> Originally I actually did want to give you all of the data back when it was all received, but Piper said uTP shouldn't have to handle that and honestly I ended up agreeing with him.

Yup, I agree now, too.

> Anyways here is my pseudocode, based on the current uTP implementation which has a pretty standard TCP-ish API. This example is assuming we are using Pipers suggested delimiter

Yeah, I think that could work. The most important thing to me is that the parsing of the `TalkRequest` and the uTP packet are happening at different layers. That's the biggest blocker I see in the current WIP:

https://github.com/ethereum/trin/pull/85/files#diff-200ac81888bc624bfdb1269b4989366d24b77ad48ef811ccb82417716e9b7fb4R377-R382

```rust
impl UtpListener {
    pub async fn process_utp_request(&mut self, talk_request: TalkRequest) {
        let packet_header = PacketHeader::decode(talk_request.body());
        let packet_type = packet_header.type_();
        let connection_id = packet_header.connection_id;
        let node_id = talk_request.node_id();
```

I'm happy to chat more after the call tomorrow if you're not sure how to separate them out.

Reply:
> Who is "you guys" here?

By "you guys" I meant the people starting the connection/using the library.

> Heh, this time I understand "you guys", but also it suggests that there are some people chiseling specs into stone. We are just a loosely assembled collective, building out a draft spec together. You're in as much as anyone. So you can use "we" instead to remind yourself that we're all just figuring it out together. :)

Oh damn, I wasn't aware of that 😄, I will keep that in mind for the future :)

> So, I'll bite: what's another way to correctly deliver an incoming utp packet, when there are two streams opened with the same peer?

![image](https://user-images.githubusercontent.com/31669092/131430028-5fe2f3c7-241f-4d82-a506-6f386e9d517e.png)

We would be fine because the ConnectionKey uses both the connection_id and the peer's id, ideally we would make only one connection peer a node, but with the current implementation we would be fine in Trin's usecase.

>Yeah, I think that could work. The most important thing to me is that the parsing of the TalkRequest and the uTP packet are happening at different layers. That's the biggest blocker I see in the current WIP:

Yeah I think you are right I will just pass in the payload Vec and the nodes ID

Reply:
Currently, timeouts aren't fully implemented in the library I still need to think of a way to add that, but I don't think this is a blocker necessarily